src/app/services/keycloak.service.ts

```diff
@@ -74,134 +63,168 @@ export class KeycloakService {
-      const queryString = `user/${userToken.sub}`;
-      this.http.get<User[]>(`${this.pathAPI}/${queryString}`, {})
-      .subscribe(
-        res => this.addUserIfNone(res, userToken),
-        err => console.log('Error retrieving user: ', err)
-      )
+  checkUser(userToken: User): Observable<User[]> {
+    if (userToken && userToken.email) {
+      let fields = [
+        '_id',
+        '_schemaName',
+        'firstName',
+        'lastName',
+        'displayName',
+        'email',
+        'project',
+        'projectPermissions',
+        'read',
+        'write',
+        'delete',
+      ].join('|');
+      const queryString = encodeURIComponent(userToken.email);
+      return this.http.get<User[]>(`${this.pathAPI}/user/email/${queryString}?fields=${fields}`, {});
     }
   }
 
   /**
-   * Checks if users is returned and adds a user entry if not
+   * Add a new user object to the DB.
    *
-   * @param userArray Array of users
-   * @param userToken Keycloak parsed token
-   * @return void
+   * @param {User} userToken The user to add.
+   * @returns {Observable[<User>]}
    */
-  addUserIfNone(userArray: User[], userToken: User): void {
-    if (userArray.length === 0) {
-      console.log('User does not yet exist, adding: ', userArray)
+  addNewUser(userToken): Observable<User> {
+    // Creating the user in the DB for the first time.
+    const user = new User({
+      idirUserGuid: userToken.idir_user_guid,
+      firstName: userToken.given_name,
+      lastName: userToken.family_name,
+      displayName: `${userToken.given_name} ${userToken.family_name}`,
+      email: userToken.email,
+      projectPermissions: []
+    });
 
-      // Creating the user in the DB for the first time.
-      // Empty project permissions.
-      const user = new User({
-        sub: userToken.sub,
-        firstName: userToken.given_name,
-        lastName: userToken.family_name,
-        displayName: `${userToken.given_name} ${userToken.family_name}`,
-        email: userToken.email,
-        projectPermissions: []
-      });
+    return this.http.post<User>(`${this.pathAPI}/user`, user, {});
+  }
 
-      this.http.post<User>(`${this.pathAPI}/user`, user, {})
-      .subscribe(
-        res => console.log('Added new user', res),
-        err => console.log('Error adding new user: ', err)
-        )
-      } else {
-        console.log('User already exists', userArray)
-      }
+  /**
+   * Takes the user's token, gets the GUID from it, and updates the user's
+   * DB entry with it.
+   *
+   * @param {User} user The user object to add to.
+   * @param {User} userToken The user token to get the GUID from.
+   * @returns {Observable<User>}
+   */
+  updateUserWithGuid(user: User, userToken): Observable<User> {
+      // Update user object with GUID.
+      user.idirUserGuid = userToken.idir_user_guid;
+      return this.http.put<User>(`${this.pathAPI}/user/${user._id}`, user, {});
   }
 
   isValidForSite() {
```

Comment:
The isValidForSite() function name seems a bit misleading, if i'm reading this correctly this is checking if you are sysadmin or not.
I guess if you need this role regardless to be in the admin, it makes sense.  But to me a sysadmin means a system administrator and not a user administrator.

Reply:
Yeah I agree with you re: "sysadmin" and what it implies. I don't know what the original intention was with calling the user a "sysadmin" but it only means that they can view and edit projects in the backend. Basically, yes, it should be "admin." However, "sysadmin" is what we've got unless we change the wording one day